File: Test-Corpus/codex-gpt54/B02_synthetic/pointer-comparison-ascii-art/translated_rust/src/shape.rs
#![allow(dead_code)]

pub const MAX_SHAPE_WIDTH: usize = 80;
pub const MAX_SHAPE_HEIGHT: usize = 30;
pub const MAX_SHAPE_NAME: usize = 32;
pub const SHAPE_COUNT: usize = 10;

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
#[repr(i32)]
pub enum ShapeType {
    Tree = 0,
    Tractor = 1,
    House = 2,
    Sun = 3,
    Cloud = 4,
    Flower = 5,
    Car = 6,
    Star = 7,
    Heart = 8,
    Rainbow = 9,
}

impl ShapeType {
    pub fn from_i32(value: i32) -> Option<Self> {
        match value {
            0 => Some(Self::Tree),
            1 => Some(Self::Tractor),
            2 => Some(Self::House),
            3 => Some(Self::Sun),
            4 => Some(Self::Cloud),
            5 => Some(Self::Flower),
            6 => Some(Self::Car),
            7 => Some(Self::Star),
            8 => Some(Self::Heart),
            9 => Some(Self::Rainbow),
            _ => None,
        }
    }
}

pub struct Shape {
    pub shape_type: ShapeType,
    pub name: &'static str,
    pub art: &'static [&'static str],
    pub width: i32,
    pub height: i32,
}

pub struct ShapeManager {
    shapes: [Box<Shape>; SHAPE_COUNT],
}

impl ShapeManager {
    pub fn new() -> Self {
        Self {
            shapes: [
                Box::new(init_tree()),
                Box::new(init_tractor()),
                Box::new(init_house()),
                Box::new(init_sun()),
                Box::new(init_cloud()),
                Box::new(init_flower()),
                Box::new(init_car()),
                Box::new(init_star()),
                Box::new(init_heart()),
                Box::new(init_rainbow()),
            ],
        }
    }

    pub fn get(&self, shape_type: ShapeType) -> &Shape {
        &self.shapes[shape_type as usize]
    }

    pub fn get_by_i32(&self, value: i32) -> Option<&Shape> {
        ShapeType::from_i32(value).map(|shape_type| self.get(shape_type))
    }
}

pub fn shape_print(shape: Option<&Shape>) {
    match shape {
        Some(shape) => {
            println!("{}:", shape.name);
            for line in shape.art.iter().take(shape.height as usize) {
                println!("{line}");
            }
        }
        None => println!("(null shape)"),
    }
}

pub fn shape_equals(s1: &Shape, s2: &Shape) -> bool {
    std::ptr::eq(s1, s2)
}

pub fn shape_type_name(value: i32) -> &'static str {
    match ShapeType::from_i32(value) {
        Some(ShapeType::Tree) => "Tree",
        Some(ShapeType::Tractor) => "Tractor",
        Some(ShapeType::House) => "House",
        Some(ShapeType::Sun) => "Sun",
        Some(ShapeType::Cloud) => "Cloud",
        Some(ShapeType::Flower) => "Flower",
        Some(ShapeType::Car) => "Car",
        Some(ShapeType::Star) => "Star",
        Some(ShapeType::Heart) => "Heart",
        Some(ShapeType::Rainbow) => "Rainbow",
        None => "Unknown",
    }
}

fn init_tree() -> Shape {
    Shape {
        shape_type: ShapeType::Tree,
        name: "Tree",
        height: 7,
        width: 11,
        art: &[
            "    /\\    ",
            "   /  \\   ",
            "  /____\\  ",
            "  /    \\  ",
            " /______\\ ",
            "    ||    ",
            "    ||    ",
        ],
    }
}

fn init_tractor() -> Shape {
    Shape {
        shape_type: ShapeType::Tractor,
        name: "Tractor",
        height: 6,
        width: 20,
        art: &[
            "      ________     ",
            "     |        |___ ",
            "     |  []  []|   |",
            "  ___|________|___|",
            " /  o        o   \\",
            "|___|        |___| ",
        ],
    }
}

fn init_house() -> Shape {
    Shape {
        shape_type: ShapeType::House,
        name: "House",
        height: 7,
        width: 13,
        art: &[
            "     /\\     ",
            "    /  \\    ",
            "   /____\\   ",
            "   |    |   ",
            "   | [] |   ",
            "   |    |   ",
            "   |____|   ",
        ],
    }
}

fn init_sun() -> Shape {
    Shape {
        shape_type: ShapeType::Sun,
        name: "Sun",
        height: 7,
        width: 11,
        art: &[
            "  \\  |  / ",
            "   \\ | /  ",
            "--- (@) ---",
            "   / | \\  ",
            "  /  |  \\ ",
            "          ",
            "          ",
        ],
    }
}

fn init_cloud() -> Shape {
    Shape {
        shape_type: ShapeType::Cloud,
        name: "Cloud",
        height: 4,
        width: 16,
        art: &[
            "   _____       ",
            "  /     \\_    ",
            " /  ___  _\\  ",
            "(__/   \\_)   ",
        ],
    }
}

fn init_flower() -> Shape {
    Shape {
        shape_type: ShapeType::Flower,
        name: "Flower",
        height: 7,
        width: 9,
        art: &[
            "  \\|/  ",
            " -(@)- ",
            "  /|\\  ",
            "   |   ",
            "   |   ",
            "  / \\  ",
            " /   \\ ",
        ],
    }
}

fn init_car() -> Shape {
    Shape {
        shape_type: ShapeType::Car,
        name: "Car",
        height: 4,
        width: 16,
        art: &[
            "  ____       ",
            " /|_||_\\____ ",
            "( o     o  ) ",
            " -----------  ",
        ],
    }
}

fn init_star() -> Shape {
    Shape {
        shape_type: ShapeType::Star,
        name: "Star",
        height: 5,
        width: 9,
        art: &[
            "    *    ",
            "   ***   ",
            "  *****  ",
            " ******* ",
            "*********",
        ],
    }
}

fn init_heart() -> Shape {
    Shape {
        shape_type: ShapeType::Heart,
        name: "Heart",
        height: 6,
        width: 11,
        art: &[
            " *** ***  ",
            "*********  ",
            "*********  ",
            " ******* ",
            "  *****  ",
            "   ***   ",
        ],
    }
}

fn init_rainbow() -> Shape {
    Shape {
        shape_type: ShapeType::Rainbow,
        name: "Rainbow",
        height: 5,
        width: 21,
        art: &[
            "      _______      ",
            "    /         \\    ",
            "   /           \\   ",
            "  /             \\  ",
            " /               \\ ",
        ],
    }
}

File: .cache/4/translated/codex-gpt56-sol/f7424ef428a79c5b88e793a692326db8b129891c5351bf5ccb368d418fc9280e/code/c_src/src/shape.c
/*
 * Copyright 2025 MIT Lincoln Laboratory
 * Permission is hereby granted, free of charge,
 * to any person obtaining a copy of this software
 * and associated documentation files (the "Software"),
 * to deal in the Software without restriction,
 * including without limitation the rights to use, copy,
 * modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software,
 * and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 * 
 * The above copyright notice and this permission notice
 * shall be included in all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO
 * THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE
 * FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE
 * OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */
// shape.c
#include "shape.h"
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <stdint.h>

// Singleton shape instances
static shape_t *shapes[SHAPE_COUNT] = {NULL};

static void init_tree(shape_t *shape) {
    shape->type = SHAPE_TREE;
    strcpy(shape->name, "Tree");
    shape->height = 7;
    shape->width = 11;
    
    strcpy(shape->art[0], "    /\\    ");
    strcpy(shape->art[1], "   /  \\   ");
    strcpy(shape->art[2], "  /____\\  ");
    strcpy(shape->art[3], "  /    \\  ");
    strcpy(shape->art[4], " /______\\ ");
    strcpy(shape->art[5], "    ||    ");
    strcpy(shape->art[6], "    ||    ");
}

static void init_tractor(shape_t *shape) {
    shape->type = SHAPE_TRACTOR;
    strcpy(shape->name, "Tractor");
    shape->height = 6;
    shape->width = 20;
    
    strcpy(shape->art[0], "      ________     ");
    strcpy(shape->art[1], "     |        |___ ");
    strcpy(shape->art[2], "     |  []  []|   |");
    strcpy(shape->art[3], "  ___|________|___|");
    strcpy(shape->art[4], " /  o        o   \\");
    strcpy(shape->art[5], "|___|        |___| ");
}

static void init_house(shape_t *shape) {
    shape->type = SHAPE_HOUSE;
    strcpy(shape->name, "House");
    shape->height = 7;
    shape->width = 13;
    
    strcpy(shape->art[0], "     /\\     ");
    strcpy(shape->art[1], "    /  \\    ");
    strcpy(shape->art[2], "   /____\\   ");
    strcpy(shape->art[3], "   |    |   ");
    strcpy(shape->art[4], "   | [] |   ");
    strcpy(shape->art[5], "   |    |   ");
    strcpy(shape->art[6], "   |____|   ");
}

static void init_sun(shape_t *shape) {
    shape->type = SHAPE_SUN;
    strcpy(shape->name, "Sun");
    shape->height = 7;
    shape->width = 11;
    
    strcpy(shape->art[0], "  \\  |  / ");
    strcpy(shape->art[1], "   \\ | /  ");
    strcpy(shape->art[2], "--- (@) ---");
    strcpy(shape->art[3], "   / | \\  ");
    strcpy(shape->art[4], "  /  |  \\ ");
    strcpy(shape->art[5], "          ");
    strcpy(shape->art[6], "          ");
}

static void init_cloud(shape_t *shape) {
    shape->type = SHAPE_CLOUD;
    strcpy(shape->name, "Cloud");
    shape->height = 4;
    shape->width = 16;
    
    strcpy(shape->art[0], "   _____       ");
    strcpy(shape->art[1], "  /     \\_    ");
    strcpy(shape->art[2], " /  ___  _\\  ");
    strcpy(shape->art[3], "(__/   \\_)   ");
}

static void init_flower(shape_t *shape) {
    shape->type = SHAPE_FLOWER;
    strcpy(shape->name, "Flower");
    shape->height = 7;
    shape->width = 9;
    
    strcpy(shape->art[0], "  \\|/  ");
    strcpy(shape->art[1], " -(@)- ");
    strcpy(shape->art[2], "  /|\\  ");
    strcpy(shape->art[3], "   |   ");
    strcpy(shape->art[4], "   |   ");
    strcpy(shape->art[5], "  / \\  ");
    strcpy(shape->art[6], " /   \\ ");
}

static void init_car(shape_t *shape) {
    shape->type = SHAPE_CAR;
    strcpy(shape->name, "Car");
    shape->height = 4;
    shape->width = 16;
    
    strcpy(shape->art[0], "  ____       ");
    strcpy(shape->art[1], " /|_||_\\____ ");
    strcpy(shape->art[2], "( o     o  ) ");
    strcpy(shape->art[3], " -----------  ");
}

static void init_star(shape_t *shape) {
    shape->type = SHAPE_STAR;
    strcpy(shape->name, "Star");
    shape->height = 5;
    shape->width = 9;
    
    strcpy(shape->art[0], "    *    ");
    strcpy(shape->art[1], "   ***   ");
    strcpy(shape->art[2], "  *****  ");
    strcpy(shape->art[3], " ******* ");
    strcpy(shape->art[4], "*********");
}

static void init_heart(shape_t *shape) {
    shape->type = SHAPE_HEART;
    strcpy(shape->name, "Heart");
    shape->height = 6;
    shape->width = 11;
    
    strcpy(shape->art[0], " *** ***  ");
    strcpy(shape->art[1], "*********  ");
    strcpy(shape->art[2], "*********  ");
    strcpy(shape->art[3], " ******* ");
    strcpy(shape->art[4], "  *****  ");
    strcpy(shape->art[5], "   ***   ");
}

static void init_rainbow(shape_t *shape) {
    shape->type = SHAPE_RAINBOW;
    strcpy(shape->name, "Rainbow");
    shape->height = 5;
    shape->width = 21;
    
    strcpy(shape->art[0], "      _______      ");
    strcpy(shape->art[1], "    /         \\    ");
    strcpy(shape->art[2], "   /           \\   ");
    strcpy(shape->art[3], "  /             \\  ");
    strcpy(shape->art[4], " /               \\ ");
}

void shape_manager_init(void) {
    // Allocate each shape once (singleton pattern)
    for (int i = 0; i < SHAPE_COUNT; i++) {
        shapes[i] = malloc(sizeof(shape_t));
        if (!shapes[i]) {
            fprintf(stderr, "Error: Failed to allocate shape\n");
            exit(1);
        }
    }
    
    // Initialize each shape
    init_tree(shapes[SHAPE_TREE]);
    init_tractor(shapes[SHAPE_TRACTOR]);
    init_house(shapes[SHAPE_HOUSE]);
    init_sun(shapes[SHAPE_SUN]);
    init_cloud(shapes[SHAPE_CLOUD]);
    init_flower(shapes[SHAPE_FLOWER]);
    init_car(shapes[SHAPE_CAR]);
    init_star(shapes[SHAPE_STAR]);
    init_heart(shapes[SHAPE_HEART]);
    init_rainbow(shapes[SHAPE_RAINBOW]);
}

void shape_manager_cleanup(void) {
    for (int i = 0; i < SHAPE_COUNT; i++) {
        free(shapes[i]);
        shapes[i] = NULL;
    }
}

shape_t* shape_get(shape_type_t type) {
    if (type < 0 || type >= SHAPE_COUNT) {
        return NULL;
    }
    return shapes[type];
}

void shape_print(const shape_t *shape) {
    if (!shape) {
        printf("(null shape)\n");
        return;
    }
    
    printf("%s:\n", shape->name);
    for (int i = 0; i < shape->height; i++) {
        printf("%s\n", shape->art[i]);
    }
}

int shape_equals(const shape_t *s1, const shape_t *s2) {
    return s1 == s2 ? 1 : 0;
}

const char* shape_type_name(shape_type_t type) {
    switch (type) {
        case SHAPE_TREE: return "Tree";
        case SHAPE_TRACTOR: return "Tractor";
        case SHAPE_HOUSE: return "House";
        case SHAPE_SUN: return "Sun";
        case SHAPE_CLOUD: return "Cloud";
        case SHAPE_FLOWER: return "Flower";
        case SHAPE_CAR: return "Car";
        case SHAPE_STAR: return "Star";
        case SHAPE_HEART: return "Heart";
        case SHAPE_RAINBOW: return "Rainbow";
        default: return "Unknown";
    }
}
